{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "76b68221-a772-4590-8454-17c1868c3857",
   "metadata": {},
   "source": [
    "# label data\n",
    "- determine biosamples related to lung and lung cancer\n",
    "- prepare ids.csv for nextflow pipeline downloading FASTQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be972f75-e10f-47e3-96ea-ccdfbe64d688",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from utils import Utils\n",
    "from label_data import LabelData\n",
    "from parse_soft import ParseSoft\n",
    "from search_soft import SearchSoft\n",
    "from enrich_soft import EnrichSoft\n",
    "from parse_sra import ParseSra\n",
    "from slicer import Slicer\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08cc9bfb-f9fc-4e1a-a845-5b01a5c1e4c1",
   "metadata": {},
   "source": [
    "## prepare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f08a3385-65ac-4a61-ac44-9c5543da6ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load pairwising data\n",
    "pmid_geo = Utils.from_json('../results/pmid_geo.json')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a841377e-1789-417c-93d8-18ad152580dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the direcotry stores soft files\n",
    "data_dir = '/home/yuan/data'\n",
    "\n",
    "# exported labeling data\n",
    "label_dir = '../data/labels'\n",
    "if not os.path.isdir(label_dir):\n",
    "    os.mkdir(label_dir)\n",
    "#\n",
    "fastq_dir = '/home/yuan/rawdata/SRX'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "00814f51-68c8-4dc7-9c66-b7477471db3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GEO': 'GSE286399',\n",
       " 'geo_http': 'https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE286399',\n",
       " 'title': 'single cell RNA-seq analysis of lung cancer cell lines treated with JTE-607',\n",
       " 'PMID': None,\n",
       " 'taxid': '9606',\n",
       " 'platform': [{'platform_id': 'GPL24676',\n",
       "   'platform_title': 'Illumina NovaSeq 6000 (Homo sapiens)',\n",
       "   'platform_technology': 'high-throughput sequencing'}],\n",
       " 'samples': {'GSM8726430': {'sample_id': 'GSM8726430',\n",
       "   'characteristics': {'tissue': 'Lung cancer cell line',\n",
       "    'cell_line': 'NCI-H1299',\n",
       "    'cell_type': 'Lung cancer cell line',\n",
       "    'treatment': 'DMSO'},\n",
       "   'scrnaseq': False,\n",
       "   'Sample_description': ['Library name: sample1'],\n",
       "   'BioSample': 'SAMN46202173',\n",
       "   'SRA': 'SRX27313887',\n",
       "   'SRA_url': 'https://www.ncbi.nlm.nih.gov/sra?term=SRX27313887',\n",
       "   'BioSample_url': 'https://www.ncbi.nlm.nih.gov/biosample/SAMN46202173'},\n",
       "  'GSM8726431': {'sample_id': 'GSM8726431',\n",
       "   'characteristics': {'tissue': 'Lung cancer cell line',\n",
       "    'cell_line': 'NCI-H1299',\n",
       "    'cell_type': 'Lung cancer cell line',\n",
       "    'treatment': 'JTE-607'},\n",
       "   'scrnaseq': False,\n",
       "   'Sample_description': ['Library name: sample2'],\n",
       "   'BioSample': 'SAMN46202172',\n",
       "   'SRA': 'SRX27313888',\n",
       "   'SRA_url': 'https://www.ncbi.nlm.nih.gov/sra?term=SRX27313888',\n",
       "   'BioSample_url': 'https://www.ncbi.nlm.nih.gov/biosample/SAMN46202172'},\n",
       "  'GSM8726432': {'sample_id': 'GSM8726432',\n",
       "   'characteristics': {'tissue': 'Lung cancer cell line',\n",
       "    'cell_line': 'NCI-H226',\n",
       "    'cell_type': 'Lung cancer cell line',\n",
       "    'treatment': 'DMSO'},\n",
       "   'scrnaseq': False,\n",
       "   'Sample_description': ['Library name: sample3'],\n",
       "   'BioSample': 'SAMN46202171',\n",
       "   'SRA': 'SRX27313889',\n",
       "   'SRA_url': 'https://www.ncbi.nlm.nih.gov/sra?term=SRX27313889',\n",
       "   'BioSample_url': 'https://www.ncbi.nlm.nih.gov/biosample/SAMN46202171'}}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test parse, and enrich\n",
    "geo ='GSE286399'\n",
    "parser = ParseSoft(data_dir)\n",
    "soft_path = parser.soft_local_path(geo)\n",
    "softer = SearchSoft(soft_path)\n",
    "data = softer.filter_data(softer.parse_rows)\n",
    "enriched = EnrichSoft(data)()\n",
    "enriched"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef05731d-6c1a-4a23-94eb-90e2c93d8953",
   "metadata": {},
   "source": [
    "## select GEO"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3beb7f9d-3ac0-4b90-a30c-211747efb905",
   "metadata": {},
   "source": [
    "### H1299"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f7b0aa1b-5858-4923-a205-8f75d742c680",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSE286399, GSE280041, GSE144357, GSE121309, GSE183590, GSE148729, \n",
      "6 GEO data are created.\n"
     ]
    }
   ],
   "source": [
    "parser = ParseSoft(data_dir)\n",
    "cl = LabelData(label_dir)\n",
    "\n",
    "n = 0\n",
    "geo_h1299 = ['GSE286399', 'GSE280041', 'GSE144357', 'GSE121309', 'GSE183590', 'GSE148729']\n",
    "for geo in geo_h1299:\n",
    "    print(geo, end=', ')\n",
    "    soft_path = parser.soft_local_path(geo)\n",
    "    softer = SearchSoft(soft_path)\n",
    "    data = softer.filter_data(softer.parse_rows)\n",
    "    enriched = EnrichSoft(data)()\n",
    "    cl.save(enriched)\n",
    "    n += 1\n",
    "print(f'\\n{n} GEO data are created.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e70e9abc-99b7-4fb4-9c0e-2ba040f6e9d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "616c871e-2ab5-4e1e-91ea-f7b8ec2d615f",
   "metadata": {},
   "source": [
    "### cell marker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a73f1df0-9d82-4c08-9f3a-4a600dbf9a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellmarker = pd.ExcelFile('../results/Cell_marker_Seq.xlsx')\n",
    "cm = cellmarker.parse('seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "783809fe-efc3-4fcf-bb94-ed4dd0d234ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[30259978 30523199 30554520 30650190 30784054 31221805 31233341 31289132\n",
      " 31299246 31333652 31405848 31834999 31840053 31892341 31996486 32004478\n",
      " 32072637 32109386 32112047 32122885 32203281 32246845 32317009 32317643\n",
      " 32373206 32398875 32405060 32497778 32580738 32603599 32810439 32822576\n",
      " 32832598 32832599 32849643 32882007 32968798 32973742 33057196 33083004\n",
      " 33123174 33144684 33178221 33377642 33382972 33500718 33514641 33571124\n",
      " 33598101 33657410 33705361 33717172 33822772 33879239 33953163 33972311\n",
      " 34017124 34030460 34049947 34247147 34313733 34330889 34475869 34504485\n",
      " 34603282 34624218 34663877 34715018 34764257 34780851 34804043 34876692\n",
      " 34914922 34916290 35078977 35108060 35126365 35184398 35213222 35216676\n",
      " 35354645 35430336]\n"
     ]
    }
   ],
   "source": [
    "# get PMID\n",
    "lung_cm = cm[(cm['species']=='Human') & cm['tissue_type'].str.contains('Lung')]\n",
    "pmid_list= np.unique(lung_cm['PMID'])\n",
    "print(pmid_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4849cbd7-43a8-43c2-84b3-caba0f416981",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSE112274, GSE121611, GSE122960, GSE128033, GSE128169, GSE133747, GSE137811, GSE137805, GSE137799, GSE124885, GSE132771, GSE147066, GSE145926, GSE140819, GSE135851, GSE135893, GSE136831, GSE217722, GSE162936, GSE162499, GSE162500, GSE162498, GSE158055, GSE166059, GSE166034, GSE166036, GSE166033, GSE166035, GSE166037, GSE168710, GSE148071, GSE164829, GSE156311, GSE168299, \n",
      "34 GEO data are created.\n"
     ]
    }
   ],
   "source": [
    "# retrieve soft data\n",
    "parser = ParseSoft(data_dir)\n",
    "cl = LabelData(label_dir)\n",
    "n = 0\n",
    "for pmid in pmid_list:\n",
    "    pmid_keys = Slicer.PMID(pmid)\n",
    "    geo_pool = Utils.key_get(pmid_geo, pmid_keys)\n",
    "    for geo in geo_pool:\n",
    "        print(geo, end=', ')\n",
    "        soft_path = parser.soft_local_path(geo)\n",
    "        softer = SearchSoft(soft_path)\n",
    "        data = softer.filter_data(softer.parse_rows)\n",
    "        enriched = EnrichSoft(data)()\n",
    "        cl.save(enriched)\n",
    "        n += 1\n",
    "print(f'\\n{n} GEO data are created.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83c5d683-b3dd-4f4e-8d9e-00737576288a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "409fd38c-8663-4cb5-8445-850d580876d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110\n"
     ]
    }
   ],
   "source": [
    "# include all PMID human related\n",
    "human_cm = cm[cm['species']=='Human']\n",
    "is_lung = human_cm.apply(lambda x: x.str.lower().str.contains('lung').any(), axis=1)\n",
    "human_cm = human_cm[is_lung]\n",
    "human_pmid_list= np.unique(human_cm['PMID'])\n",
    "print(len(human_pmid_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "23691f2d-9562-4e8f-9207-4956f42e04e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSE112274, GSE121611, GSE122960, GSE127472, GSE127462, GSE127813, GSE127471, GSE130148, GSE128033, GSE128169, GSE133747, GSE137811, GSE137805, GSE137799, GSE124885, GSE132771, GSE147066, GSE125188, GSE145926, GSE140819, GSE135851, GSE135893, GSE136831, GSE158127, GSE161089, GSE217722, GSE155249, GSE162936, GSE162499, GSE162500, GSE162498, GSE158055, GSE166059, GSE166034, GSE166036, GSE166033, GSE166035, GSE166037, GSE168710, GSE148071, GSE164829, GSE156311, GSE155515, GSE180864, GSE185044, GSE185043, GSE185045, GSE154826, GSE168299, GSE190510, GSE196303, GSE180908, GSE180063, \n",
      "53 GEO data are created.\n"
     ]
    }
   ],
   "source": [
    "# retrieve soft data\n",
    "parser = ParseSoft(data_dir)\n",
    "cl = LabelData(label_dir)\n",
    "n = 0\n",
    "for pmid in human_pmid_list:\n",
    "    pmid_keys = Slicer.PMID(pmid)\n",
    "    geo_pool = Utils.key_get(pmid_geo, pmid_keys)\n",
    "    for geo in geo_pool:\n",
    "        print(geo, end=', ')\n",
    "        soft_path = parser.soft_local_path(geo)\n",
    "        softer = SearchSoft(soft_path)\n",
    "        data = softer.filter_data(softer.parse_rows)\n",
    "        enriched = EnrichSoft(data)()\n",
    "        cl.save(enriched)\n",
    "        n += 1\n",
    "print(f'\\n{n} GEO data are created.')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce726170-59c6-4af2-baab-d6a945d05472",
   "metadata": {},
   "source": [
    "## parse SRR given BioSample and SRX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5757d760-d1a3-4ae2-a804-36ab0e6298f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data are determined previously\n",
    "samn_srr = Utils.from_json('../results/samn_srr.json')\n",
    "srr_fastq = Utils.from_json('../results/srr_fastq_urls.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4847497a-3d54-484a-98f5-2fd782ab17ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given BioSample, parse SRR accessions if they exist\n",
    "for data in Utils.json_iter(label_dir):\n",
    "    data = ParseSra.parse_srr(data, samn_srr)\n",
    "    cl.save(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59a73af6-f733-42fb-9a8b-7ad1b6328733",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start GSE133747...\t0 out of 8 SRR are updated.\n",
      "Start GSE128033...\t0 out of 0 SRR are updated.\n",
      "Start GSE128169...\t0 out of 0 SRR are updated.\n",
      "Start GSE154826...\t0 out of 249 SRR are updated.\n",
      "Start GSE112274...\t0 out of 2721 SRR are updated.\n",
      "Start GSE280041...\t0 out of 155 SRR are updated.\n",
      "Start GSE168710...\t0 out of 0 SRR are updated.\n",
      "Start GSE168299...\t0 out of 7 SRR are updated.\n",
      "Start GSE140819...\t0 out of 0 SRR are updated.\n",
      "Start GSE166059...\t0 out of 92 SRR are updated.\n",
      "Start GSE166034...\t0 out of 13 SRR are updated.\n",
      "Start GSE166033...\t0 out of 9 SRR are updated.\n",
      "Start GSE166037...\t0 out of 4 SRR are updated.\n",
      "Start GSE166035...\t0 out of 14 SRR are updated.\n",
      "Start GSE166036...\t0 out of 48 SRR are updated.\n",
      "Start GSE145926...\t0 out of 20 SRR are updated.\n",
      "Start GSE183590...\t0 out of 1455 SRR are updated.\n",
      "Start GSE125188...\t0 out of 8 SRR are updated.\n",
      "Start GSE217722...\t0 out of 0 SRR are updated.\n",
      "Start GSE135851...\t0 out of 7 SRR are updated.\n",
      "Start GSE135893...\t0 out of 227 SRR are updated.\n",
      "Start GSE130148...\t0 out of 0 SRR are updated.\n",
      "Start GSE286399...\t0 out of 3 SRR are updated.\n",
      "Start GSE148071...\t0 out of 0 SRR are updated.\n",
      "Start GSE148729...\t0 out of 175 SRR are updated.\n",
      "Start GSE127813...\t0 out of 0 SRR are updated.\n",
      "Start GSE127472...\t0 out of 0 SRR are updated.\n",
      "Start GSE127462...\t0 out of 0 SRR are updated.\n",
      "Start GSE127471...\t0 out of 0 SRR are updated.\n",
      "Start GSE122960...\t0 out of 0 SRR are updated.\n",
      "Start GSE190510...\t0 out of 28 SRR are updated.\n",
      "Start GSE121309...\t0 out of 12 SRR are updated.\n",
      "Start GSE121611...\t0 out of 1 SRR are updated.\n",
      "Start GSE156311...\t0 out of 0 SRR are updated.\n",
      "Start GSE196303...\t0 out of 0 SRR are updated.\n",
      "Start GSE132771...\t0 out of 68 SRR are updated.\n",
      "Start GSE180063...\t0 out of 44 SRR are updated.\n",
      "Start GSE180864...\t0 out of 0 SRR are updated.\n",
      "Start GSE180908...\t0 out of 19 SRR are updated.\n",
      "Start GSE144357...\t"
     ]
    }
   ],
   "source": [
    "# retrieve FTP URLs of *gz files given their SRR accessions\n",
    "for data in Utils.json_iter(label_dir):\n",
    "    data = ParseSra.parse_ftp_fastq(data, srr_fastq)\n",
    "    cl.save(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bdff262-6902-4d2f-9c99-2eae9aaf7c44",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0cb0070-6934-4adb-92d6-d982e5f2d550",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "539e8b63-d319-4558-8cd7-72b24df5cc9d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8df38b7-fa30-4912-bf74-7ec270009029",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f32de545-ff5a-4eeb-bc76-c7f04d8876f4",
   "metadata": {},
   "source": [
    "## select biosamples and "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "e4a53b5a-9ff0-4a68-ae31-bac738f189a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>cd34+_hsc_donor</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>disease_state</th>\n",
       "      <th>fetal_lung_tissue_donor</th>\n",
       "      <th>genotype/variation</th>\n",
       "      <th>group</th>\n",
       "      <th>inoculation_dose</th>\n",
       "      <th>library_type</th>\n",
       "      <th>...</th>\n",
       "      <th>mouse_strain</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>samplename_old</th>\n",
       "      <th>sex</th>\n",
       "      <th>technology</th>\n",
       "      <th>time_point</th>\n",
       "      <th>tissue</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3926539</th>\n",
       "      <td>9 weeks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3926540</th>\n",
       "      <td>9 weeks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3926541</th>\n",
       "      <td>9 weeks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3926542</th>\n",
       "      <td>9 weeks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3926543</th>\n",
       "      <td>3.5 months</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4889468</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>airway basal stem cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X Genomics Single Cell RNA Sequencing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cigarette smoke</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3557941</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD45+CD14-CD16-CD3E+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3557942</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD45+CD14-CD16-CD3E+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3557943</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD45+CD14-CD16-CD3E-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4968655</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RUES2 human embryonic stem cell</td>\n",
       "      <td>RUES2-derived lung cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>day50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>399 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   age cd34+_hsc_donor                        cell_line  \\\n",
       "GSM3926539     9 weeks             NaN                              NaN   \n",
       "GSM3926540     9 weeks             NaN                              NaN   \n",
       "GSM3926541     9 weeks             NaN                              NaN   \n",
       "GSM3926542     9 weeks             NaN                              NaN   \n",
       "GSM3926543  3.5 months             NaN                              NaN   \n",
       "...                ...             ...                              ...   \n",
       "GSM4889468         NaN             NaN                              NaN   \n",
       "GSM3557941         NaN             NaN                              NaN   \n",
       "GSM3557942         NaN             NaN                              NaN   \n",
       "GSM3557943         NaN             NaN                              NaN   \n",
       "GSM4968655         NaN             NaN  RUES2 human embryonic stem cell   \n",
       "\n",
       "                           cell_type disease_state fetal_lung_tissue_donor  \\\n",
       "GSM3926539                       NaN           NaN                     NaN   \n",
       "GSM3926540                       NaN           NaN                     NaN   \n",
       "GSM3926541                       NaN           NaN                     NaN   \n",
       "GSM3926542                       NaN           NaN                     NaN   \n",
       "GSM3926543                       NaN           NaN                     NaN   \n",
       "...                              ...           ...                     ...   \n",
       "GSM4889468   airway basal stem cells           NaN                     NaN   \n",
       "GSM3557941      CD45+CD14-CD16-CD3E+           NaN                     NaN   \n",
       "GSM3557942      CD45+CD14-CD16-CD3E+           NaN                     NaN   \n",
       "GSM3557943      CD45+CD14-CD16-CD3E-           NaN                     NaN   \n",
       "GSM4968655  RUES2-derived lung cells           NaN                     NaN   \n",
       "\n",
       "           genotype/variation group inoculation_dose library_type  ...  \\\n",
       "GSM3926539                NaN   NaN              NaN          NaN  ...   \n",
       "GSM3926540                NaN   NaN              NaN          NaN  ...   \n",
       "GSM3926541                NaN   NaN              NaN          NaN  ...   \n",
       "GSM3926542                NaN   NaN              NaN          NaN  ...   \n",
       "GSM3926543                NaN   NaN              NaN          NaN  ...   \n",
       "...                       ...   ...              ...          ...  ...   \n",
       "GSM4889468                NaN   NaN              NaN          NaN  ...   \n",
       "GSM3557941                NaN   NaN              NaN          NaN  ...   \n",
       "GSM3557942                NaN   NaN              NaN          NaN  ...   \n",
       "GSM3557943                NaN   NaN              NaN          NaN  ...   \n",
       "GSM4968655                NaN   NaN              NaN          NaN  ...   \n",
       "\n",
       "           mouse_strain patient_id sample_type samplename_old     sex  \\\n",
       "GSM3926539          NaN        NaN         NaN            NaN    male   \n",
       "GSM3926540          NaN        NaN         NaN            NaN  female   \n",
       "GSM3926541          NaN        NaN         NaN            NaN    male   \n",
       "GSM3926542          NaN        NaN         NaN            NaN  female   \n",
       "GSM3926543          NaN        NaN         NaN            NaN    male   \n",
       "...                 ...        ...         ...            ...     ...   \n",
       "GSM4889468          NaN        NaN         NaN            NaN     NaN   \n",
       "GSM3557941          NaN        NaN         NaN            NaN     NaN   \n",
       "GSM3557942          NaN        NaN         NaN            NaN     NaN   \n",
       "GSM3557943          NaN        NaN         NaN            NaN     NaN   \n",
       "GSM4968655          NaN        NaN         NaN            NaN     NaN   \n",
       "\n",
       "                                         technology time_point tissue  \\\n",
       "GSM3926539                                      NaN        NaN    NaN   \n",
       "GSM3926540                                      NaN        NaN    NaN   \n",
       "GSM3926541                                      NaN        NaN    NaN   \n",
       "GSM3926542                                      NaN        NaN    NaN   \n",
       "GSM3926543                                      NaN        NaN    NaN   \n",
       "...                                             ...        ...    ...   \n",
       "GSM4889468  10X Genomics Single Cell RNA Sequencing        NaN    NaN   \n",
       "GSM3557941                                      NaN        NaN    NaN   \n",
       "GSM3557942                                      NaN        NaN    NaN   \n",
       "GSM3557943                                      NaN        NaN    NaN   \n",
       "GSM4968655                                      NaN      day50    NaN   \n",
       "\n",
       "           tissue_type        treatment  \n",
       "GSM3926539         NaN              NaN  \n",
       "GSM3926540         NaN              NaN  \n",
       "GSM3926541         NaN              NaN  \n",
       "GSM3926542         NaN              NaN  \n",
       "GSM3926543         NaN              NaN  \n",
       "...                ...              ...  \n",
       "GSM4889468         NaN  cigarette smoke  \n",
       "GSM3557941         NaN              NaN  \n",
       "GSM3557942         NaN              NaN  \n",
       "GSM3557943         NaN              NaN  \n",
       "GSM4968655         NaN              NaN  \n",
       "\n",
       "[399 rows x 23 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# confirm characteristics of biosample\n",
    "features, sample_srr = {}, {}\n",
    "cl = LabelData(label_dir)\n",
    "for data in cl.json_iter():\n",
    "    if data['taxid'] == '9606':\n",
    "        for sample_id, sample in data['samples'].items():\n",
    "            if sample.get('scrnaseq') and sample['SRR']:\n",
    "                #print(data['GEO'], sample_id)\n",
    "                features[sample_id] = sample['characteristics']\n",
    "                sample_srr[sample_id] = sample['SRR']\n",
    "#\n",
    "df = pd.DataFrame.from_dict(features).T\n",
    "df = df[sorted(list(df))]\n",
    "df.to_csv('../data/features.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1b1f07b-5fb5-4994-bc3c-6bb469d60ea3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d391c1c5-0eba-48a5-a308-c1169920320a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52 400\n"
     ]
    }
   ],
   "source": [
    "# prepare SRR id list for download\n",
    "num_geo = num_sample = 0\n",
    "cl = LabelData(label_dir)\n",
    "with open('../results/srr_ids.txt', 'w') as f:\n",
    "    for data in cl.json_iter():\n",
    "        if data['taxid'] == '9606':\n",
    "            num_geo += 1\n",
    "            for sample_id, sample in data['samples'].items():\n",
    "                if sample.get('scrnaseq') and sample['SRR']:\n",
    "                    #print(data['GEO'], sample_id)\n",
    "                    num_sample += 1\n",
    "                    for srr_acc, local_path in sample['SRR'].items():\n",
    "                        if not local_path:\n",
    "                            # print(f\"fastq-dump {srr_acc} -O /home/yuan/results/fastq --gzip\")\n",
    "                            f.write(srr_acc + '\\n')\n",
    "print(num_geo, num_sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e58a4f0f-1693-499a-bdd0-a9d03c376a64",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "713c5641-3f6c-4794-832c-651a16f8b9f4",
   "metadata": {},
   "source": [
    "## manual label"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3f40bbd-dacb-402b-9deb-4b275ed261a6",
   "metadata": {},
   "source": [
    "### lung cancer key=cell_line\n",
    "- NCI-H1299 is an epithelial-like cell that was isolated from the lung of a White, 43-year-old, male patient with carcinoma.\n",
    "- A549 cells were isolated from the lung tissue of a White, male, lung cancer patient.\n",
    "- Calu-3 epithelial cells are isolated from lung tissue derived from a 25-year-old, White, male patient with lung adenocarcinoma who received prior therapy with cytoxan, bleomycin, and adriamycin. This cell line is valuable for SARS-CoV-2 propagation in vitro, is a suitable transfection host, and has applications in cancer and toxicology research."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ab8f048e-4137-4caa-82d2-b343a44d7de3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H1299 40\n",
      "Calu3 90\n",
      "A549 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>fastq_1</th>\n",
       "      <th>fastq_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1299_SRR11549938</td>\n",
       "      <td>/home/yuan/data/SRA/SRR115/038/SRR11549938.fas...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H1299_SRR11549939</td>\n",
       "      <td>/home/yuan/data/SRA/SRR115/039/SRR11549939.fas...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H1299_SRR11549940</td>\n",
       "      <td>/home/yuan/data/SRA/SRR115/040/SRR11549940.fas...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H1299_SRR11549941</td>\n",
       "      <td>/home/yuan/data/SRA/SRR115/041/SRR11549941.fas...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H1299_SRR11549942</td>\n",
       "      <td>/home/yuan/data/SRA/SRR115/042/SRR11549942.fas...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample                                            fastq_1  \\\n",
       "0  H1299_SRR11549938  /home/yuan/data/SRA/SRR115/038/SRR11549938.fas...   \n",
       "1  H1299_SRR11549939  /home/yuan/data/SRA/SRR115/039/SRR11549939.fas...   \n",
       "2  H1299_SRR11549940  /home/yuan/data/SRA/SRR115/040/SRR11549940.fas...   \n",
       "3  H1299_SRR11549941  /home/yuan/data/SRA/SRR115/041/SRR11549941.fas...   \n",
       "4  H1299_SRR11549942  /home/yuan/data/SRA/SRR115/042/SRR11549942.fas...   \n",
       "\n",
       "  fastq_2  \n",
       "0    None  \n",
       "1    None  \n",
       "2    None  \n",
       "3    None  \n",
       "4    None  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def lung_cancer_cell_line(df):\n",
    "    res = []\n",
    "    for name in ['H1299', 'Calu3', 'A549']:\n",
    "        df1 = df[df['cell_line']==name]\n",
    "        print(name, len(df1))\n",
    "        for sample_id in list(df1.index):\n",
    "            for srr_acc, fq in sample_srr[sample_id].items():\n",
    "                if fq:\n",
    "                    label, fq1, fq2 = f\"{name}_{srr_acc}\", fq.get('R1'), fq.get('R2')\n",
    "                    res.append((label, fq1, fq2))\n",
    "    res = pd.DataFrame(res, columns=['sample', 'fastq_1', 'fastq_2'])\n",
    "    return res\n",
    "df1 = lung_cancer_cell_line(df)\n",
    "df1.to_csv('../data/samplesheet_cellline_lungcancer.csv', index=False)\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "190b0e98-79b1-4795-936d-5db52a7dc890",
   "metadata": {},
   "source": [
    "### lung cell line\n",
    "- WI-38 cell line is the first human diploid cell line to be used in human vaccine preparation. WI-38 cells were isolated from the lung tissue of a 3-month-old, female, embryo. WI-38 is used in viruscide testing.\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3de17d50-22c9-4349-be8c-67a414ad4902",
   "metadata": {},
   "source": [
    "### colon cancer, key=cell_line\n",
    "- HCT 116 is an adherent cell line isolated from the colon of a patient with colon cancer. It has a mutation in codon 13 of the ras-proto-oncogene. This cell line is near-diploid and has a relatively stable genetic profile, making it a valuable in vitro model. This line can be utilized in cancer research and gastrointestinal (GI) research.\n",
    "- Caco-2 [Caco2] is an adherent cell line isolated from colon tissue derived from a patient with colorectal adenocarcinoma. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fac56aaa-ee55-4738-b063-e2a059be3209",
   "metadata": {},
   "source": [
    "### normal cell line\n",
    "- cell_line=\"RUES2 human embryonic stem cell\". RUES2 is a human embryonic stem cell line derived from a de-identified frozen embryo that was originally generated for reproductive purposes.\n",
    "- cell_type = \"airway basal stem cells\"\n",
    "- cell_type = \"CD45+\", group=\"healthy control\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "65d548e8-c09b-4c69-963a-cc417d829e4f",
   "metadata": {},
   "source": [
    "### lung tissue\n",
    "Normal\n",
    "- tissue=lung, tissue_type=Normal\n",
    "- tissue=lung, group=healthy control\n",
    "- tissue = Peripheral blood mononuclear cell, disease_state=Health donors\n",
    "patient\n",
    "- tissue=lung, tissue_type=Tumor\n",
    "- tissue=lung, disease_state = Coronavirus infected disease-19 (COVID-19)\n",
    "- tissue=lung, group = severe COVID-19 patient\n",
    "- tissue=lung, disease_state = Lymphangioleiomyomatosis (LAM) patient\n",
    "- tissue = Peripheral blood mononuclear cell, disease_state=Patient with ASS-ILD"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
